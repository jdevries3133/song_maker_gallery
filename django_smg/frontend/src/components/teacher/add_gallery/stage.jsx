import React, { useState } from "react";
import PropTypes from "prop-types";

import { StagedGroup } from "./snippets";
import { ErrorArray } from "Common/custom_error";
import styled, {
  H2,
  H3,
  P,
  Input,
  Textarea,
  Button,
  Div,
  Description,
} from "Styles";

import { SizeLimitCounter } from "./verify";

export const TITLE_LENGTH_LIMIT = 100;

const LargeButton = styled(Button)`
  padding: 50px;
  background-color: #ffc107;
  border: 3px solid black;
`;

export const Stage = (props) => {
  const [blanket, setBlanket] = useState(null);
  const submitValidation = () => {
    if (props.titleValue.length >= TITLE_LENGTH_LIMIT) {
      setBlanket(
        <ErrorArray
          header="Title Too Long"
          message={[
            `Gallery title must be less than ${TITLE_LENGTH_LIMIT} characters ` +
              `long. Your title is currently ${props.titleValue.length} ` +
              `characters long.`,
          ]}
          onOk={() => setBlanket(null)}
        />
      );
    } else {
      props.confirmCreate();
    }
  };
  return (
    <Div>
      {blanket}
      <Description>
        <H2>Your Staged Gallery</H2>
        <Div>
          <H3>Gallery Name:</H3>
          <Input
            data-testid="titleInput"
            placeholder="Name of your gallery here"
            value={props.titleValue}
            onChange={(e) => props.titleInput(e)}
          />{" "}
          <SizeLimitCounter
            length={props.titleValue.length}
            limit={TITLE_LENGTH_LIMIT}
            warnLimit={TITLE_LENGTH_LIMIT - 10}
            hideUntil={50}
          />
          <H3>Gallery Description:</H3>
          <P>
            You may use this default description, or change it to whatever you
            prefer.
          </P>
          <Textarea
            value={props.descriptionValue}
            onChange={(e) => props.descriptionInput(e)}
          />
          <H3>Staged Goups</H3>
          {props.groups.map((group) => (
            <StagedGroup
              key={group.join("") + Math.random().toString()}
              unstageGroupHandler={props.unstageGroupHandler}
              group={group}
            />
          ))}
        </Div>
        <LargeButton data-testid="submit" onClick={submitValidation}>
          <P>Create Gallery</P>
          <P>
            (Or, add another group by uploading another <code>.csv</code> file.)
          </P>
        </LargeButton>
      </Description>
    </Div>
  );
};

Stage.propTypes = {
  titleValue: PropTypes.string.isRequired,
  titleInput: PropTypes.func.isRequired,
  descriptionValue: PropTypes.string.isRequired,
  descriptionInput: PropTypes.func.isRequired,
  groups: PropTypes.array.isRequired,
  unstageGroupHandler: PropTypes.func.isRequired,
  confirmCreate: PropTypes.func.isRequired,
};
