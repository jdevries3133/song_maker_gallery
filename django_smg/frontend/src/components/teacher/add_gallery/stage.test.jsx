import React from "react";
import renderer from "react-test-renderer";
import { render, cleanup, fireEvent, act } from "@testing-library/react";
import "@testing-library/jest-dom";
import "jest-styled-components";

import { Stage, TITLE_LENGTH_LIMIT } from "./stage";

const MOCK_STAGED_INIT = {
  titleValue: "title",
  titleInput: jest.fn(),
  descriptionValue: "description",
  descriptionInput: jest.fn(),
  groups: [[]],
  unstageGroupHandler: jest.fn(),
  confirmCreate: jest.fn(),
};

let mockStaged;

beforeEach(() => {
  mockStaged = { ...MOCK_STAGED_INIT };
});

afterEach(cleanup);

/* Helper function for rendering Stage with title of n characters */
const renderWithChars = (numChars) => {
  return {
    ...render(<Stage {...mockStaged} titleValue={"c".repeat(numChars)} />),
  };
};

describe("stage", () => {
  it("matches snapshot", () => {
    const tree = renderer.create(<Stage {...mockStaged} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("has a counter that shows current num chars entered", () => {
    const { getByTestId } = renderWithChars(60);
    expect(getByTestId("titleLenLimit")).toHaveTextContent(
      `60/${TITLE_LENGTH_LIMIT}`
    );
  });

  // TODO: toHaveStyle assertions are broken
  // https://github.com/styled-components/styled-components/issues/3297
  //
  // Restore these texts when they are fixed

  // it("shows counter after 50 chars are entered", () => {
  //   const testForCharCounts = [0, 10, 20, 30, 49];
  //   testForCharCounts.forEach((count) => {
  //     const { queryByText } = renderWithChars(count);
  //     // counter is not in the DOM
  //     expect(queryByText(`${count}/${TITLE_LENGTH_LIMIT}`)).toHaveStyle(
  //       "display: none;"
  //     );
  //   });
  // });

  // it("turns counter orange after limit - 10 exceeded", () => {
  //   const { getByTestId } = renderWithChars(TITLE_LENGTH_LIMIT - 10);
  //   expect(getByTestId("titleLenLimit")).toHaveStyle("color: orange;");
  // });

  // it("turns counter red after limit exceeded", () => {
  //   const { getByTestId } = renderWithChars(TITLE_LENGTH_LIMIT);
  //   expect(getByTestId("titleLenLimit")).toHaveStyle("color: rgb(228, 16, 0);");
  // });

  it("prevents submission of title longer than 100 chars", async () => {
    const { getByTestId, getByText } = renderWithChars(101);
    act(() => {
      fireEvent.click(getByTestId("submit"));
    });
    expect(getByTestId("errorArray header")).toHaveTextContent(
      "Title Too Long"
    );
    expect(
      getByText(`must be less than ${TITLE_LENGTH_LIMIT}`, { exact: false })
    ).toBeTruthy();
    expect(
      getByText(`Your title is currently 101 characters long`, {
        exact: false,
      })
    ).toBeTruthy();
  });

  it("calls confirmCreate for valid data", async () => {
    const { getByTestId } = render(<Stage {...mockStaged} />);
    act(() => {
      fireEvent.click(getByTestId("submit"));
    });
    expect(mockStaged.confirmCreate).toHaveBeenCalledTimes(1);
  });
});
