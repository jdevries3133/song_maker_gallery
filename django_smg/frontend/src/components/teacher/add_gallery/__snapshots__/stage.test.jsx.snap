// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`stage matches snapshot 1`] = `
.c0 {
  text-align: center;
  padding: inherit;
}

.c1 {
  display: inline-block;
  background-color: "#88ff00";
  font-weight: 400;
  padding: 30px;
  border-radius: 20px;
}

.c4 {
  text-align: center;
  padding: 0 20px 0 20px;
}

.c8 {
  display: inline-block;
  font-size: 1rem;
  font-weight: bold;
  padding: 20px;
  border-radius: 20px;
  cursor: pointer;
  background-color: salmon;
}

.c8:hover {
  background-color: white;
}

.c10 {
  display: inline-block;
  font-size: 1rem;
  font-weight: bold;
  padding: 20px;
  border-radius: 20px;
  cursor: pointer;
  background-color: #f7943e;
}

.c10:hover {
  background-color: white;
}

.c2 {
  font-size: 20px;
  padding: 10px;
  border-radius: 5px;
}

.c2[type="checkbox"] {
  width: 30px;
  height: 30px;
}

.c7 {
  -webkit-transform: rotate(180deg);
  -ms-transform: rotate(180deg);
  transform: rotate(180deg);
}

.c6 {
  margin: auto;
  margin-top: 15px;
}

.c9 {
  margin: auto;
  margin-top: 15px;
  text-align: center;
}

.c3 {
  padding-left: "10px",  display:none;
}

.c11 {
  padding: 50px;
  background-color: #ffc107;
  border: 3px solid black;
}

.c5 {
  font-family: Helvetica,Arial,sans-serif;
  font-size: 19px;
  line-height: 1.5;
  margin: 0px;
  width: 85%;
  height: 263px;
  box-shadow: 1px 1px 1px #999;
  border: 1px solid #ccc;
  border-radius: 5px;
}

@media (max-width:400px) {
  .c8 {
    padding: 10px;
  }
}

@media (max-width:350px) {
  .c8 {
    padding: 5px;
  }
}

@media (max-width:400px) {
  .c10 {
    padding: 10px;
  }
}

@media (max-width:350px) {
  .c10 {
    padding: 5px;
  }
}

<div
  className="c0"
>
  <h2
    className="c1"
  >
    Your Staged Gallery
  </h2>
  <div
    className="c0"
  >
    <h3
      className="styles__H3-kZrHQh"
    >
      Gallery Name:
    </h3>
    <input
      className="c2"
      data-testid="titleInput"
      onChange={[Function]}
      placeholder="Name of your gallery here"
      value="title"
    />
     
    <span
      className="c3"
      data-testid="titleLenLimit"
    >
      5
      /
      100
    </span>
    <h3
      className="styles__H3-kZrHQh"
    >
      Gallery Description:
    </h3>
    <p
      className="c4"
    >
      You may use this default description, or change it to whatever you prefer.
    </p>
    <textarea
      className="c5"
      onChange={[Function]}
      value="description"
    />
    <h3
      className="styles__H3-kZrHQh"
    >
      Staged Goups
    </h3>
    <table
      className="c6"
    >
      <tbody>
        <tr>
          <td>
            <button
              className="c7"
              data-testid="sgToggleDropdown"
              onClick={[Function]}
            >
              â—€
            </button>
          </td>
          <td
            style={
              Object {
                "fontSize": "24px",
              }
            }
          />
          <td>
            <button
              className="c8"
              color="salmon"
              onClick={[Function]}
            >
              Unstage Group
            </button>
          </td>
        </tr>
      </tbody>
    </table>
    <table
      className="c9"
    >
      <tbody />
    </table>
  </div>
  <button
    className="c10 c11"
    data-testid="submit"
    onClick={[Function]}
  >
    <p
      className="c4"
    >
      Create Gallery
    </p>
    <p
      className="c4"
    >
      (Or, add another group by uploading another 
      <code>
        .csv
      </code>
       file.)
    </p>
  </button>
</div>
`;
